src/NUnitEngine/nunit.engine/Runners/WorkItemTracker.cs

```diff
@@ -0,0 +1,113 @@
+    /// even forced cancellation or kill.
+    /// 
+    /// The engine needs to protect itself from such problems by assuring that
+    /// any test for which cancellation is requested is fully cancelled, with
+    /// nothing left behind. Further, it needs to generate notifications to
+    /// tell the runner what happened.
+    /// 
+    /// WorkItemTracker examines test events and keeps track of those tests
+    /// that have started but not yet finished. It implements ITestEventListener
+    /// in order to capture events. It allows waiting for all items to complete.
+    /// Once the test has been cancelled, it provide notifications to the runner
+    /// so the information may be displayed.
+    /// </summary>
+    internal class WorkItemTracker : ITestEventListener
+    {
+        private List<XmlNode> _itemsInProcess = new List<XmlNode>();
+        private ManualResetEvent _allItemsComplete = new ManualResetEvent(false);
+        private object _trackerLock = new object();
+        
+        public void Clear()
+        {
+            lock (_trackerLock)
+            {
+                _itemsInProcess.Clear();
+                _allItemsComplete.Reset();
+            }
+        }
+
+        public bool WaitForCompletion(int millisecondsTimeout)
+        {
+            return _allItemsComplete.WaitOne(millisecondsTimeout);
+        }
+
+        public void IssuePendingNotifications(ITestEventListener listener)
+        {
+            lock (_trackerLock)
+            {
+                int count = _itemsInProcess.Count;
+
+                // Signal completion of all pending suites, in reverse order
+                while (count > 0)
+                    listener.OnTestEvent(CreateNotification(_itemsInProcess[--count]));
+            }
+        }
+
+        private static string CreateNotification(XmlNode startElement)
+        {
+            bool isSuite = startElement.Name == "start-suite";
+
+            XmlNode notification = XmlHelper.CreateTopLevelElement(isSuite ? "test-suite" : "test-case");
+            if (isSuite)
+                notification.AddAttribute("type", startElement.GetAttribute("type"));
+            notification.AddAttribute("id", startElement.GetAttribute("id"));
+            notification.AddAttribute("name", startElement.GetAttribute("name"));
+            notification.AddAttribute("fullname", startElement.GetAttribute("fullname"));
+            notification.AddAttribute("result", "Failed");
+            notification.AddAttribute("label", "Cancelled");
+            XmlNode failure = notification.AddElement("failure");
+            XmlNode message = failure.AddElementWithCDataSection("message", "Test run cancelled by user");
+            return notification.OuterXml;
+        }
+
+        void ITestEventListener.OnTestEvent(string report)
+        {
+            XmlNode xmlNode = XmlHelper.CreateXmlNode(report);
+
+            lock (_trackerLock)
+            {
```

Comment:
This is going to be high pretty throughput code, I'm thinking. Would it be more efficient to use a multi-thread safe collection type, and just lock it when we give start to force the cancellation messages?

Reply:
In the GUI I followed my usual practice... assume no performance problem and only fix it if it shows up. Haven't seen one yet.